editor/api/sidebar.js

```diff
@@ -0,0 +1,107 @@
+/* eslint no-console: [ 'error', { allow: [ 'error' ] } ] */
+
+/* External dependencies */
+import { isFunction } from 'lodash';
+
+/* Internal dependencies */
+import store from '../store';
+import { setActivePanel } from '../store/actions';
+import { applyFilters } from '@wordpress/hooks';
+
+const sidebars = {};
+
+/**
+ * Registers a sidebar with the editor.
+ *
+ * A button will be shown in the settings menu to open the sidebar. The sidebar
+ * can be manually opened by calling the `activateSidebar` function.
+ *
+ * @param {string} name              The name of the sidebar. Should be in
+ *                                   `[plugin]/[sidebar]` format.
+ * @param {Object}   settings        The settings for this sidebar.
+ * @param {string}   settings.title  The name to show in the settings menu.
+ * @param {Function} settings.render The function that renders the sidebar.
+ *
+ * @returns {Object} The final sidebar settings object.
+ */
+export function registerSidebar( name, settings ) {
+	settings = {
+		name,
+		...settings,
+	};
+
+	if ( typeof name !== 'string' ) {
+		console.error(
+			'Sidebar names must be strings.'
+		);
+		return;
+	}
+	if ( ! /^[a-z][a-z0-9-]*\/[a-z][a-z0-9-]*$/.test( name ) ) {
+		console.error(
+			'Sidebar names must contain a namespace prefix, include only lowercase alphanumeric characters or dashes, and start with a letter. Example: my-plugin/my-custom-sidebar'
+		);
+		return;
+	}
+	if ( ! settings || ! isFunction( settings.render ) ) {
+		console.error(
+			'The "render" property must be specified and must be a valid function.'
+		);
+		return;
+	}
+	if ( sidebars[ name ] ) {
+		console.error(
+			'Sidebar "' + name + '" is already registered.'
+		);
+	}
+
+	if ( ! ( 'title' in settings ) || settings.title === '' ) {
+		console.error(
+			'The sidebar "' + name + '" must have a title.'
+		);
+		return;
+	}
+	if ( typeof settings.title !== 'string' ) {
+		console.error(
+			'Sidebar titles must be strings.'
+		);
+		return;
+	}
+
+	settings = applyFilters( 'editor.registerSidebar', settings, name );
+
+	return sidebars[ name ] = settings;
+}
+
+/**
+ * Retrieves the sidebar settings object.
+ *
+ * @param {string} name The name of the sidebar to retrieve.
+ *
+ * @returns {false|Object} The settings object of the sidebar. Or false if the
```

Comment:
The "return `false` if not exists" pattern never quite made sense to me in the context of JavaScript, since we already have two concepts to represent an empty/unset value: `null` and `undefined`. I'd rather we return one of these instead, preferably `null` since we already do this in [a number of selectors](https://github.com/WordPress/gutenberg/blob/master/editor/store/selectors.js) and it semantically better represents the "explicitly empty" use-case (vs. unset).

Aside: If we did return false, the return value type would be Boolean.